MultiRenderTarget extended support

Orikson opened the pull request:
Minimally added support for MultiRenderTarget binding 2D, 2D Array, Cube Map, and (soon) 3D textures to different draw buffers.

From the following [forum post](https://forum.babylonjs.com/t/multirendertarget-with-2d-array-textures/37185).

Additional considerations from the forum:
* MRT w/ 3D textures
* MRT w/ Cube map array textures (WebGPU)
* Hardware maximum draw buffers (`gl.MAX_DRAW_BUFFERS`)

bjsplat commented:
Snapshot stored with reference name:
refs/pull/13435/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13435/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13435/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/13435/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/13435/merge
https://gui.babylonjs.com/?snapshot=refs/pull/13435/merge
https://nme.babylonjs.com/?snapshot=refs/pull/13435/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/13435/merge#BCU1XR#0

Popov72 commented on packages/dev/core/src/Materials/Textures/multiRenderTarget.ts:
```diff
@@ -290,6 +317,37 @@ export class MultiRenderTarget extends RenderTargetTexture {
+        }
+
+        const renderTarget = this.renderTarget as WebGLRenderTargetWrapper;
+        renderTarget.setTexture(this.textures[index].getInternalTexture()!, index, false, faceIndexOrLayer);
```
Replace these two lines with a call to `this.renderTarget.setLayerAndFaceIndex()`.

Popov72 reviewed (changes requested):
Good job!

The main change is to split `faceIndexOrLayer` into `faceIndex` and `layerIndex`, because we definitely need to take into account the support of cube map arrays in WebGPU.

You also need to update `Engines/RenderTargetWrapper._cloneRenderTargetWrapper` to take into account the new options in the multi case.

Also, we will need to add a test to our visualization test suite => you can do this at the end, once we agree with the PR.

I will do the WebGPU port when the PR is accepted and merged for WebGL.

Orikson commented:
Sorry for the long wait, been super busy with school.

As you suggested, I split `faceIndexOrLayer` into `faceIndex` and `layerIndex` where necessary (`IMultiRenderTargetOptions`, `MultiRenderTarget`, `RenderTargetWrapper`, etc.). I also fixed `_cloneRenderTargetWrapper`. 

I also added spaces for `TEXTURE_CUBE_MAP_ARRAY` in the code, but wasn't so sure what I should be checking for (2D array has `.is2DArray`, cube map has `.isCube`). I just left it commented out as `.isCubeArray`.

Updated usage looks like
```typescript
let mrt = new BABYLON.MultiRenderTarget(
    "mrt", 
    { width: 32, height: 32 }, 
    4, // number of draw buffers
    scene, 
    { 
        types: new Array(3).fill(BABYLON.Constants.TEXTURETYPE_FLOAT),
        samplingModes: new Array(3).fill(BABYLON.Constants.TEXTURE_NEAREST_SAMPLINGMODE),
        targetTypes: [BABYLON.Constants.TEXTURE_2D_ARRAY, BABYLON.Constants.TEXTURE_CUBE_MAP, BABYLON.Constants.TEXTURE_2D],
        faceIndex: [0, 3, 0, 0],
        layerIndex: [0, 0, 0, 0],
        layerCounts: [32, 0, 0, 32]
    }
);

// Sets the attachment at index 3 to the 0th texture in the list
mrt.setInternalTexture(mrt.textures[0].getInternalTexture(), 3);
mrt.setLayerAndFaceIndex(3, 1, 0);
```

As for the type forcing in the 3D texture case, I was messing around with different formats because I still can't seem to get `framebufferTextureLayer` to bind a 3D texture. The other three texture types work great though.

Popov72 commented:
Good job!

I tested the changes locally, and noticed things were a bit more complicated because we have to handle MSAA properly, as well as make sure that the recreation of the multi render target object works as expected, which means that if we bind two layers of a 2DArrayTexture, for eg, after the recreation, we should not end up with two new 2DArrayTexture but only one which is bound to two different attachments, just like we had before the recreation...

I thought it was easier if I made the changes myself, and it allowed me to do extensive testing and find things I wouldn't have found by just looking at the existing code. So, here's the diff between the current PR and my changes:

https://github.com/Popov72/Babylon.js/compare/a51d6943e04c463dd33cdfb9f6da3161b67d169e...b7fe107e526d3a6dd2da3c5be101b3b313223576

Feel free to ask any questions wrt to the changes (I don't guarantee I did not make any mistake)!

The PG I used for my tests: http://localhost:1338/#409YEN#18
=> note that the Inspector does not work correctly for 2DArrayTexture, we need to add correct support

Note also that rendering to a 3D texture is actually bugged in Angle (there are some issues in their repo). It seems to work when using native OpenGL, but since it is not the default backend in Chrome, I did not use 3DTexture in my PG.

Popov72 commented:
@Orikson Any news about this PR?

Orikson commented:
Changes makes sense. Looks like clone works correctly now. I guess the good thing about using `layerCounts` internally is that we don't have to keep it up to date (when cloning/setting internal textures).

On line 110 of `webGLRenderTargetWrapper.ts` I think if the default `faceIndexOrLayer` is 0 then this breaks backwards compatibility, particularly for legacy calls to `setInternalTexture`. Line 113 (below) would force WebGL to bind the 0th face of a cube map instead of a 2d texture. 

```typescript
const target = faceIndexOrLayer !== -1 ? gl.TEXTURE_CUBE_MAP_POSITIVE_X + faceIndexOrLayer : gl.TEXTURE_2D;
```

Line 175 of `webGLRenderTargetWrapper.ts` should probably bind `this.faceIndices[index]` instead of `this.layerIndices[index]`. 

As an aside, is it worth considering adding an additional parameter to modify the depth/stencil texture bound to the framebuffer? For example, one use case might be having a 2D array depth texture where each layer corresponds to a texture in the MRT.

Popov72 commented:
> On line 110 of webGLRenderTargetWrapper.ts I think if the default faceIndexOrLayer is 0 then this breaks backwards compatibility, particularly for calls to setInternalTexture.

Good catch, it should be `faceIndexOrLayer = faceIndexOrLayer ?? -1;` instead.

> Line 175 of webGLRenderTargetWrapper.ts should probably bind this.faceIndices[index] instead of this.layerIndices[index].

Indeed.

> As an aside, is it worth considering adding an additional parameter to modify the depth/stencil texture bound to the framebuffer? For example, one use case might be having a 2D array depth texture where each layer corresponds to a texture in the MRT.

I'm not sure for this one, I can't really see a use case for this? Also, I am concerned that this will add complexities for a niche case.... What do you think @sebavan?

Also, if that is ok for you, can you merge my changes to your PR so we can flag it ready for review? Note that you will also need to add a visualization test for it.

Thanks a lot!

Orikson commented:
Merged your changes into my PR and fixed the things mentioned before.

Kind of stuck on the visualization test -- do I make a playground showcasing changes (e.g. the one you created or the one I created before) and add 

```json
{
    "title": "MultiRenderTarget with different texture types",
    "playgroundId": "#PLAYGROUND",
    "excludedEngines": ["webgpu"]
}, 
```

and then in the playground test cloning and binding of different layers/faces/textures?

sebavan commented:
yup it would be the way to go :-)

bjsplat commented:
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13435/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13435/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

Popov72 commented:
@Orikson Can you resolve the conflicts so that we can restart the CI?

Thanks in advance!

Popov72 commented:
There are some formatting problems in `packages/dev/core/src/Engines/constants.ts` and `packages/dev/inspector/src/components/actionTabs/tabs/propertyGrids/materials/texturePropertyGridComponent.tsx`: you can run the "Format Document With..." command in VSCode and choose "Prettier - Code formatter" to fix them on each file.

bjsplat commented:
Visualization tests for webgl2 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13435/merge/testResults/webgl2/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13435/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

bjsplat commented:
Visualization tests for webgl1 have failed. If some tests failed because the snapshots do not match, the report can be found at

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/13435/merge/testResults/webgl1/index.html

If tests were successful afterwards, this report might not be available anymore.

Popov72 commented:
Thanks for the update!

There's one file to remove from the PR, but I think the PR is now ready for others to review!

cc @sebavan

sebavan commented on packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:
```diff
@@ -75,23 +75,46 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
+        if (this._engine.webGLVersion > 1) {
+            const gl = this._context as WebGL2RenderingContext;
+
+            const attachment = (<any>gl)["COLOR_ATTACHMENT" + attachmentIndex];
```
I wonder if splitting this function per type (2dArray/3d, cube, others) would not make sense it becomes quite complex what happens inside also this is only called in setFace... which already checks the different cases.

Orikson replied:
I'm worried about calls to `setTexture`, although I agree in every other case it is redundant

sebavan commented:
@Orikson any update on the feedback ?

Orikson commented on packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:
```diff
@@ -75,23 +75,46 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
+            faceIndexOrLayer = faceIndexOrLayer ?? -1;
+
+            const attachment = (<any>gl)["COLOR_ATTACHMENT" + attachmentIndex + "_WEBGL"];
+            const target = faceIndexOrLayer !== -1 ? gl.TEXTURE_CUBE_MAP_POSITIVE_X + faceIndexOrLayer : gl.TEXTURE_2D;
```
`setsetLayerAndFaceIndex` defaults to -1, so I made changes there too; I think I left it like this because it was backwards compatible. However, I don't think `_bindTextureRenderTarget` was called previously with the `faceIndex` parameter set, and nothing else relies on `layers` being set to -1, so I changed everything to optional parameters.

Popov72 commented on packages/dev/core/src/Engines/WebGL/webGLRenderTargetWrapper.ts:
```diff
@@ -165,7 +163,7 @@ export class WebGLRenderTargetWrapper extends RenderTargetWrapper {
      * @param face The face of the texture to be set (make negative to not modify)
      */
-    public setLayerAndFaceIndex(index: number = 0, layer: number = -1, face: number = -1): void {
+    public setLayerAndFaceIndex(index: number = 0, layer?: number, face?: number): void {
```
The parameter documentation should be updated accordingly.
